File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case13_streamrelate/Case02_ConnectOnEventTime.java
package com.yw.flink.example.javacases.case13_streamrelate;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.CoProcessFunction;
import org.apache.flink.util.Collector;

import java.time.Duration;

/**
 * Flink - Connect 两流合并，测试watermark
 */
public class Case02_ConnectOnEventTime {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //设置并行度为1
        env.setParallelism(1);

        //准备A流 ,002,182,183,fail,2000,20
        SingleOutputStreamOperator<StationLog> dsA = env.socketTextStream("node5", 8888)
                .map((MapFunction<String, StationLog>) line -> {
                    String[] arr = line.split(",");
                    return new StationLog(arr[0], arr[1], arr[2], arr[3], Long.valueOf(arr[4]), Long.valueOf(arr[5]));
                });
        //对A流设置watermark
        SingleOutputStreamOperator<StationLog> dsAWithWatermark = dsA
                .assignTimestampsAndWatermarks(WatermarkStrategy.<StationLog>forBoundedOutOfOrderness(Duration.ofSeconds(2))
                        .withTimestampAssigner((SerializableTimestampAssigner<StationLog>) (stationLog, l) -> stationLog.callTime));


        //准备B流 ,002,2000
        SingleOutputStreamOperator<String> dsB = env.socketTextStream("node5", 9999);
        //对A流设置watermark
        SingleOutputStreamOperator<String> dsBWithWatermark = dsB
                .assignTimestampsAndWatermarks(WatermarkStrategy.<String>forBoundedOutOfOrderness(Duration.ofSeconds(2))
                        .withTimestampAssigner((SerializableTimestampAssigner<String>) (str, l) -> Long.parseLong(str.split(",")[1])));

        //两流进行union
        dsAWithWatermark.connect(dsBWithWatermark)
                .process(new CoProcessFunction<StationLog, String, String>() {
                    //处理第一个流数据
                    @Override
                    public void processElement1(StationLog stationLog,
                                                CoProcessFunction<StationLog, String, String>.Context context,
                                                Collector<String> collector) throws Exception {
                        collector.collect("A流数据：" + stationLog.toString() + ",watermark = " + context.timerService().currentWatermark());
                    }

                    //处理第二个流中数据
                    @Override
                    public void processElement2(String str,
                                                CoProcessFunction<StationLog, String, String>.Context context,
                                                Collector<String> collector) throws Exception {
                        collector.collect("B流数据：" + str + "，watermark = " + context.timerService().currentWatermark());
                    }
                }).print();
        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case13_streamrelate/Case02_ConnectOnEventTime.scala
package com.yw.flink.example.scalacases.case13_streamrelate

import com.yw.flink.example.StationLog
import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.streaming.api.functions.co.CoProcessFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.util.Collector

import java.time.Duration

/**
  * Flink - Connect 两流合并，测试watermark
  */
object Case02_ConnectOnEventTime {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    import org.apache.flink.streaming.api.scala._

    //创建A流,001,181,182,busy,1000,10
    val ADS: DataStream[StationLog] = env.socketTextStream("node5", 8888)
      .map(line => {
        val arr: Array[String] = line.split(",")
        StationLog(arr(0), arr(1), arr(2), arr(3), arr(4).toLong, arr(5).toLong)
      })
    //设置watermark
    val adsWithWatermark: DataStream[StationLog] = ADS.assignTimestampsAndWatermarks(
      WatermarkStrategy.forBoundedOutOfOrderness[StationLog](Duration.ofSeconds(2))
        .withTimestampAssigner(new SerializableTimestampAssigner[StationLog] {
          override def extractTimestamp(t: StationLog, l: Long): Long = t.callTime
        })
    )


    //创建B流,001,1000
    val BDS: DataStream[String] = env.socketTextStream("node5", 9999)
    //设置watermark
    val bdsWithWatermark: DataStream[String] = BDS.assignTimestampsAndWatermarks(
      WatermarkStrategy.forBoundedOutOfOrderness[String](Duration.ofSeconds(2))
        .withTimestampAssigner(new SerializableTimestampAssigner[String] {
          override def extractTimestamp(t: String, l: Long): Long = t.split(",")(1).toLong
        })
    )

    //两流进行合并
    adsWithWatermark.connect(bdsWithWatermark)
      .process(new CoProcessFunction[StationLog, String, String] {
        //处理A流
        override def processElement1(in1: StationLog,
                                     context: CoProcessFunction[StationLog, String, String]#Context,
                                     collector: Collector[String]): Unit = {
          collector.collect("A流数据：" + in1.toString + ",watermark = " + context.timerService().currentWatermark())
        }

        //处理B流
        override def processElement2(in2: String,
                                     context: CoProcessFunction[StationLog, String, String]#Context,
                                     collector: Collector[String]): Unit = {

          collector.collect("B流数据：" + in2 + ",watermark = " + context.timerService().currentWatermark())
        }
      }).print()

    env.execute()
  }

}
